blaze-client/src/main/scala/org/http4s/client/blaze/PoolManager.scala

```diff
@@ -87,19 +94,28 @@ private final class PoolManager(builder: ConnectionBuilder,
             case Some(Waiting(_, callback)) =>
               logger.debug(s"Fulfilling waiting connection request: ${stats}")
               callback(stage.right)
-            case None =>
+            case None if waitQueue.isEmpty =>
               logger.debug(s"Returning idle connection to pool: ${stats}")
               idleQueue.enqueue(stage)
+
+            case None => // this connection didn't match any pending request, kill it and start a new one for a queued request
```

Comment:
After I pushed the commit I realized that I should have explicitly mentioned that invariant that either the idle queue should be empty or the wait queue should be empty. I might throw that in as a doc string or perhaps an assert in the next PR.